src/pcre2_xclass.c

```diff
@@ -441,4 +441,96 @@ while (TRUE)
   }
 }
 
+
+
+/*************************************************
+*       Match character against an ECLASS        *
+*************************************************/
+
+/* This function is called to match a character against an extended class
+used for describing characters using boolean operations on sets.
+
+Arguments:
+  c           the character
+  data_start  points to the start of the ECLASS data
+  data_end    points one-past-the-last of the ECLASS data
+  utf         TRUE if in UTF mode
+
+Returns:      TRUE if character matches, else FALSE
+*/
+
+BOOL
+PRIV(eclass)(uint32_t c, PCRE2_SPTR data_start, PCRE2_SPTR data_end, BOOL utf)
+{
+PCRE2_SPTR ptr = data_start;
+uint32_t stack = 0;
+
+/* Do a little loop, until we reach the end of the ECLASS. */
+while (ptr < data_end)
+  {
+  switch (*ptr)
+    {
+    case OP_ECLASS_OR:
+    ++ptr;
+    stack = (stack >> 1) | (stack & (uint32_t)1u);
+    break;
+
+    case OP_ECLASS_AND:
+    ++ptr;
+    stack = (stack >> 1) & (stack | ~(uint32_t)1u);
+    break;
+
+    case OP_ECLASS_SUB:
+    ++ptr;
+    stack = (stack >> 1) & (~stack | ~(uint32_t)1u);
+    break;
+
+    case OP_ECLASS_NOT:
+    ++ptr;
+    stack ^= (uint32_t)1u;
+    break;
+
+    case OP_CLASS:
+    case OP_NCLASS:
+      {
+      uint32_t matched;
+      if (c > 255)
+        matched = *ptr == OP_NCLASS;
+      else
+        matched = (((const unsigned char *)(ptr+1))[c/8] & (1u << (c&7))) != 0;
+
+      ptr += 1 + (32 / sizeof(PCRE2_UCHAR));
+      stack = (stack << 1) | matched;
+      break;
+      }
+
+#ifdef SUPPORT_WIDE_CHARS
+    case OP_XCLASS:
+      {
+      uint32_t matched = PRIV(xclass)(c, ptr + 1 + LINK_SIZE, utf);
+
+      ptr += GET(ptr, 1);
+      stack = (stack << 1) | matched;
+      break;
+      }
+#endif
+
+    case OP_ALLANY:
+    ++ptr;
+    stack = (stack << 1) | 1u;
+    break;
+
+    /* This should never occur, but compilers may mutter if there is no
+    default. */
+
+    default:
+    PCRE2_DEBUG_UNREACHABLE();
```

Comment:
I just copied it from the existing code above.

I'm happy with return statements after DEBUG_UNREACHABLE() assertions, because in release builds, the assertions aren't emitted at all, so _something_ has to happen if a customer actually triggers this condition.

Regardless of whether compilers complain, I complain if I see switch statements without a default condition! An assertion is highly beneficial for maintainability.

Reply:
fair enough, but then the argument still stands for using a `PCRE2_UNREACHABLE()` instead as that return should never happen in  release build anyway.

Reply:
Currently it looks like the macros emit:

|---|---|---|
| | Debug | Non-debug |
|---|---|---|
| PCRE2_DEBUG_UNREACHABLE | __assume(0) | do {} while (0) |
| PCRE2_UNREACHABLE | __assume(0) | __assume(0) |

This looks correct to me.

The DEBUG_UNREACHABLE version should have error-recovery code, such as a return statement, but the UNREACHABLE() version shouldn't. I like that.

I think __assume(0) is correct for MSVC.

Reply:
Actually, I think it should emit the following instead:

| | Debug | Non-debug |
|---|---|---|
| PCRE2_DEBUG_UNREACHABLE | PCRE2_ASSERT(FALSE) | do {} while (0) |
| PCRE2_UNREACHABLE | PCRE2_ASSERT(FALSE) | __assume(0) |

We want some actual runtime code to call abort() when we're in a debug build.

I can make a PR to do that separately.

Reply:
I have noticed recently that `assert()` built-in function has  pretty nice output, looks better than `abort()`